src/core/image.cpp

```diff
@@ -8345,16 +8399,69 @@ void Image::ApplyCTFPhaseFlip(CTF ctf_to_apply) {
     }
 }
 
-void Image::ApplyCTF(CTF ctf_to_apply, bool absolute, bool apply_beam_tilt, bool apply_envelope) {
+void Image::ApplyCTF(CTF const& ctf_to_apply, bool absolute, bool apply_beam_tilt, bool apply_envelope) {
     MyDebugAssertTrue(is_in_memory, "Memory not allocated");
     MyDebugAssertTrue(is_in_real_space == false, "image not in Fourier space");
     MyDebugAssertTrue(logical_z_dimension == 1, "Volumes not supported");
 
-    int j;
-    int i;
-
     long pixel_counter = 0;
 
+#if defined(MKL)
+    std::size_t batch = physical_upper_bound_complex_x + 1;
+
+    float* y_coord           = static_cast<float*>(mkl_malloc(sizeof(float) * batch, 64));
+    float* x_coord           = static_cast<float*>(mkl_malloc(sizeof(float) * batch, 64));
+    float* azimuth           = static_cast<float*>(mkl_malloc(sizeof(float) * batch, 64));
+    float* frequency_squared = static_cast<float*>(mkl_malloc(sizeof(float) * batch, 64));
+    float* buffer            = static_cast<float*>(mkl_malloc(sizeof(float) * batch, 64));
+    float* ctf_value         = static_cast<float*>(mkl_malloc(sizeof(float) * batch, 64));
+
+    for ( int j = 0; j <= physical_upper_bound_complex_y; ++j ) {
```

Comment:
To avoid using too much intermediate memory, we loop over y, and vectorize over the inner dimension x